src/main/java/com/microsoft/sqlserver/jdbc/SQLServerFMTQuery.java

```diff
@@ -0,0 +1,104 @@
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.stream.Collectors;
+
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.Token;
+
+class SQLServerFMTQuery {
+
+    private static final String FMT_ON = "SET FMTONLY ON;";
+    private static final String SELECT = "SELECT ";
+    private static final String FROM = " FROM ";
+    private static final String FMT_OFF = ";SET FMTONLY OFF;";
+    
+    private String prefix = "";
+    private ArrayList<? extends Token> tokenList = null;
+    private List<String> userColumns = new ArrayList<>();
+    private List<String> tableTarget = new ArrayList<>();
+    private List<String> possibleAliases = new ArrayList<>();
+    private List<List<String>> valuesList = new ArrayList<>();
+
+    List<String> getColumns() {
+        return userColumns;
+    }
+
+    List<String> getTableTarget() {
+        return tableTarget;
+    }
+
+    List<List<String>> getValuesList() {
+        return valuesList;
+    }
+
+    List<String> getAliases() {
+        return possibleAliases;
+    }
+
+    String constructColumnTargets() {
+        if (userColumns.contains("?")) {
+            return userColumns.stream().filter(s -> !s.equalsIgnoreCase("?")).map(s -> s.equals("") ? "NULL" : s)
+                    .collect(Collectors.joining(","));
+        } else {
+            return userColumns.isEmpty() ? "*" : userColumns.stream().map(s -> s.equals("") ? "NULL" : s)
+                    .collect(Collectors.joining(","));
+        }
+    }
+
+    String constructTableTargets() {
+        return tableTarget.stream().distinct().filter(s -> !possibleAliases.contains(s))
+                .collect(Collectors.joining(","));
+    }
+
+    String getFMTQuery() {
+        StringBuilder sb = new StringBuilder(FMT_ON);
+        if (prefix != "") {
+            sb.append(prefix);
+        }
+        sb.append(SELECT);
+        sb.append(constructColumnTargets());
+        if (!tableTarget.isEmpty()) {
+            sb.append(FROM);
+            sb.append(constructTableTargets());
+        }
+        sb.append(FMT_OFF);
+        return sb.toString();
+    }
+
+    // Do not allow default instantiation, class must be used with sql query
+    @SuppressWarnings("unused")
+    private SQLServerFMTQuery() {};
+
+    SQLServerFMTQuery(String userSql) throws SQLServerException {
+        if (null == userSql || userSql.length() == 0) {
+            SQLServerException.makeFromDriverError(null, this, SQLServerResource.getResource("R_noTokensFoundInUserQuery"), "", false);
+        }
+        InputStream stream = new ByteArrayInputStream(userSql.getBytes(StandardCharsets.UTF_8));

```

Comment:
I couldn't figure this out from documentation. Does Lexer expect a UTF8 stream?

Reply:
It's not documented that we have to use UTF-8, but most of the examples do this UTF-8 conversion and I haven't come across issues setting it this way.